### ISSUE TITLE
AKS: Unable to create cluster with underscore in name

### ISSUE THREAD

**Issue by cpinjani** [rancher/dashboard#9416]
**Setup**
- Rancher version: `v2.7-67422772955d5830b237fe84dbf1a26d266b55c1-head`
- Browser type & version: Chrome Latest

**Describe the bug**
AKS: Unable to create cluster with underscore in name. The same is valid as per AKS Docs

**To Reproduce**
Create AKS cluster via Rancher UI with "_" in name eg. rancher_test

**Expected Result**
User must be able to create AKS cluster with underscore in name as it is supported
![image](https://github.com/rancher/dashboard/assets/73099870/fddec180-2065-4f97-8f93-33a5fb999782)

**Screenshots**
![image](https://github.com/rancher/dashboard/assets/73099870/e03bf807-a5d3-47b9-be63-b7c0b306831a)

cc: @rancher/highlander

**Comment by cpinjani** [comment]
Validated on build: `v2.9-6d87a11ea46b7571646d7c3d7af704584c39fd62-head`
Seems not fixed since unable to create cluster with underscore in name.

![image](https://github.com/rancher/dashboard/assets/73099870/d04df06a-0b20-4eb4-878b-c411ed7f7867)

**Comment by yonasberhe23** [comment]
@mantis-toboggan-md I'm getting an error when the cluster name has an underscore. can you take a look please?


![Image](https://github.com/rancher/dashboard/assets/127343932/7bd8f63d-efe1-4467-bcdf-c6df0dc187c1)

**Comment by valaparthvi** [comment]
I can reproduce this issue as well on Rancher (v2.9-d2fa8a7d6116c48a814ee7983af522d614461e49-head), Dashboard (master 526f329).
cc: @mjura @yiannistri

**Comment by gaktive** [comment]
@kkaempf this may be backend related, can you have someone take a look?

**Comment by gaktive** [comment]
At this time, this may be a release blocker.

**Comment by yiannistri** [comment]
Sharing my findings regarding this issue:

- The error seems to originate from a [k8s validation library](https://github.com/kubernetes/apimachinery/blob/v0.30.2/pkg/util/validation/validation.go#L187) that checks whether a _label_ is valid. In this case, it has been repurposed to check the cluster name (unsure why). 
- The Rancher API appears to perform this check [when a new cluster is being created](https://github.com/rancher/rancher/blob/f7c67319712fe7f6130d32a8e79e94cd78425a56/pkg/api/norman/store/node/store.go#L178) regardless of the cluster driver being used. So the reported issue is not specific to AKS clusters but also affects EKS and RKE2 clusters too.
- In fact, we have UI validation that prevents underscores when creating an RKE2 cluster (as shown in the screenshot below). 

Given that this issue is not specific to AKS (or any other hosted provider) and the fact that our UI prevents users from using underscores in other parts of the app, I'm not sure how we should proceed. It feels like changing this logic would affect cluster creation in general (so not exactly a quick fix before a major release) and it would make the app somewhat inconsistent in terms of cluster naming rules.

cc @kkaempf @gaktive @mjura  for input 

![Screenshot from 2024-07-12 16-20-32](https://github.com/user-attachments/assets/08ac8563-3c2c-462b-9f66-cad41619f8d7)

**Comment by gaktive** [comment]
@nwmac based on this library and where we also don't permit underscores, it sounds like we can't easily make this work with the time we have left. Since this has been around for a while, should we tackle this later or what should we try in the current timeframe?

**Comment by nwmac** [comment]
@gaktive I don't think we allow the underscore because Rancher uses that name to name the cluster resource - this is a k8s object and I don't believe `_` is valid for that. So, AWS may allow it, but because of how Rancher works, we don't.

This would require backend changes that may not be straight forward.

Thus, we should not do this for 2.9.0 and should hand over to the backend team to own.

**Comment by nwmac** [comment]
@cpinjani Does this work with 2.8? (We migrated the UI from the old Ember to Vue in 2.9, so want to check if this is a side-effect of that).

**Comment by cpinjani** [comment]
@nwmac Its a legacy issue initially reported on v2.7 and reproducible across Rancher versions.

**Comment by gaktive** [comment]
I'll push this to 2.10.0 then. I'll put a release note label here though I think we may have this known issue about `_` in names documented somewhere already. @btat can keep me honest.

**Comment by gaktive** [comment]
Need to push this out and resync with Highlander on what the backend should be doing for validation.
